src/types/impl-trait.md

```diff
@@ -111,10 +133,193 @@ fn foo() -> impl Trait {
+    println!("{:?}", my_struct.bar);
+}
+fn new() -> MyStruct {
+    MyStruct {
+        bar: 42
+    }
+}
+```
+
+### Usage in trait impls
+
+Since type-alias-impl-trait can be referenced anywhere a type alias could be, this also means you can use them in `impl` blocks:
+
+```rust,ignore
+type Foo = impl Trait;
+
+impl Bar for Foo {}
+```
+
+There's a huge caveat though: now it's possible for there to be an impl for an opaque type *and* its hidden type:
+
+```rust,ignore
+type Foo = impl Trait;
+fn foo() -> Foo {}
+
+impl Bar for Foo {}
+impl Bar for () {} // ERROR conflicts with `impl Bar for Foo`
+```
+
+This check is *not* done by revealing the hidden type, but by checking whether a type could be a hidden type for that specific opaque type. So the following program is legal:
+
+```rust
+#![feature(type_alias_impl_trait)]
+trait Trait {}
+impl Trait for () {}
+type Foo = impl Trait;
+fn foo() -> Foo {}
+
+trait Bar {}
+impl Bar for Foo {}
+impl Bar for i32 {}
+```
+
+This is legal, because `i32` could not possibly be a hidden type of `Foo`, because it doesn't implement `Trait` which is a requirement for all hypothetical hidden types of `Foo`.
+
+> Note that this only works within the crate that defined the trait and the hidden type, as otherwise adding any trait impl would become a semver breaking change,
+> as a dependent crate could have implemented the trait for one of its own type-alias-impl-trait whose hidden type could have gotten an impl in the crate defining the trait.
+
+### Associated types
+
+Associated types can also use `impl Trait`:
+
+```rust,ignore
+impl Deref for MyType {
+    type Target = impl Trait;
+    fn deref(&self) -> &Self::Target {
+        &self.field
+    }
+}
+```
+
+While this example is fairly artificial, the real benefit is when you have unnameable types like `async` blocks:
+
+```rust,ignore
+impl IntoFuture for MyType {
+    type Output = ();
+    type Future = impl Future<Output = ()>;
+    fn into_future(self) -> Self::Future {
+        async move {
+            // do stuff here
+        }
+    }
+}
+```
+
+This way you do not need to write burdensome `Future` impls yourself. Similarly with complex `Iterator` implementations.
+
+### Defining scope
+
+Similar to return-position-impl-trait, you can only bind a hidden type of a type-alias-impl-trait within a specific "scope" (henceforth called "defining scope").
+The defining scope of a return-position-impl-trait is the function's body, excluding other items nested within that function's body.
+
+The defining scope of a type-alias-impl-trait is the scope in which it was defined. So usually a module and all its child items, but it can also be a function body, const initializer and similar scopes that can define items.
+
+Any use of the type-alias-impl-trait within the defining scope will become a **defining use** (meaning it binds a hidden type), if the type is coerced to or from, equated with, or subtyped with any other concrete type.
```

Comment:
Can this maybe add an example that illustrates the consequence of a defining scope of tait? I think there is a connection that I am missing from binding of the hidden type in a defining scope to the limitations now implied of using that type alias outside of that defining scope. (Or perhaps those limitations are the same within and without the defining scope, I'm not clear.)